reader.go

```diff
@@ -122,76 +122,56 @@ func (r *Reader) Config() ReaderConfig {
 // messages from it.
 func (r *Reader) Close() error {
 	r.mutex.Lock()
-	if !r.closed {
-		close(r.done)
-	}
+	closed := r.closed
 	r.closed = true
-	r.dirty = false
-	r.cancel()
 	r.mutex.Unlock()
+
+	r.cancel()
+	r.join.Wait()
```

Comment:
I agree, I didn't say it wouldn't be clean, I feel it increases complexity for no clear benefits. There's always something more we can test but at some point we gain very little value out of it. Here the synchronization is pretty trivial, there's only one spot where goroutines are being started, I think it's not worth adding internal abstractions.